File: src/bin/day02.rs
extern crate nom;
use nom::{
    character::complete::alpha1,
    character::complete::anychar,
    character::complete::char,
    character::complete::digit1,
    bytes::complete::tag,
    combinator::eof,
    sequence::preceded,
    sequence::tuple,
    Finish,
    IResult,
};

use std::fs;

#[derive(Debug)]
struct Line {
    min: usize,
    max: usize,
    c: char,
    pw: String,
}

fn parse_line(input: &str) -> IResult<&str, Line> {
    // 12-14 s: ssskrssssssfsxpsqsp
    let (input, (min_str, max_str, c, pw)) = tuple((
        digit1,
        preceded(char('-'), digit1),
        preceded(char(' '), anychar),
        preceded(tag(": "), alpha1),
    ))(input)?;

    let (input, _) = eof(input)?;

    Ok((
        input,
        Line {
            min: min_str.parse::<usize>().unwrap(),
            max: max_str.parse::<usize>().unwrap(),
            c,
            pw: String::from(pw),
        },
    ))
}

fn is_valid1(line: &Line) -> bool {
    let num = line.pw.chars().filter(|c| *c == line.c).count();
    num >= line.min && num <= line.max
}

fn is_valid2(line: &Line) -> bool {
    let s = line.pw.as_bytes();
    let match1 = s[line.min - 1] == line.c as u8;
    let match2 = s[line.max - 1] == line.c as u8;
    (match1 && !match2) || (match2 && !match1)
}

fn main() {
    let contents = fs::read_to_string("day02.txt").expect("Something went wrong reading the file");

    let lines: Vec<Line> = contents
        .lines()
        .map(|s| parse_line(s).finish().map(|t| t.1).unwrap())
        .collect();

        let valid1 = lines.iter().filter(|ln| is_valid1(ln)).count();
        let valid2 = lines.iter().filter(|ln| is_valid2(ln)).count();

    println!("{} passwords, {} valid in part 1", lines.len(), valid1);
    println!("{} passwords, {} valid in part 2", lines.len(), valid2);
}

File: julia/day02.jl

input = open("../inputs/day02.txt") do f
    read(f, String)
end

struct Record
    from::Int64
    to::Int64
    chr::Char
    pwd::String
end

format_re = r"^(\d+)-(\d+) ([a-z]): ([a-z]+)$"

function parseLine(ln)
    # 1-3 a: abcde
    capts = match(format_re, ln).captures
    return Record(
        parse(Int64, capts[1]),
        parse(Int64, capts[2]),
        capts[3][1],
        capts[4],
    )

end

function count_valid(is_valid::Function, records::Array{Record})::Int64
    return sum([
        1 for record in records if is_valid(record)
    ])
end

records = [ parseLine(ln) for ln in split(input, "\n") ]

valid1 = count_valid(records) do record
    matches = count(c -> c == record.chr, record.pwd)
    record.from <= matches <= record.to
end

valid2 = count_valid(records) do record
    ok1 = record.pwd[record.from] == record.chr
    ok2 = record.pwd[record.to] == record.chr
    ok1 != ok2
end

println("Part 1: $(valid1)")
println("Part 2: $(valid2)")
